Use Remote Settings for fetching studies.

rhelmer opened the pull request:
Fixes issue #296, prereq for issue #132

Checklist for reviewer:

- [ ] The description should reference a bug or github issue, if relevant.
- [ ] There must be a [`CHANGELOG.md`](./CHANGELOG.md) entry for any non-test change.
- [ ] Any change to the NPM commands must be carefully reviewed to make sure it won't break the Add-ons pipeline.
- [ ] Any version increase must follow the [release process](./docs/RELEASE_PROCESS.md).

rhelmer commented:
@Dexterp37 OK so this works (well outside of the listener part but I am confident we can figure that out).

However, looking at this now it occurs to me that we could do this in unprivileged code by just fetching the appropriate JSON directly from the Remote Settings CDN and implement the listener to fire on a timer that only runs when the user is idle.

The overhead of using the built-in Firefox API doesn't seem worth it, the only advantage is really that it supports push but that doesn't really seem very compelling (some clients can't use push and fall back to an RS timer, anyway).

@Dexterp37 wdyt?

rhelmer commented:
> @Dexterp37 OK so this works (well outside of the listener part but I am confident we can figure that out).
> 
> However, looking at this now it occurs to me that we could do this in unprivileged code by just fetching the appropriate JSON directly from the Remote Settings CDN and implement the listener to fire on a timer that only runs when the user is idle.

We may need to use https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/alarms if we go this route. It might mean another permission, but there are some restrictions on the web content APIs per that doc.

Dexterp37 commented on core-addon/Core.js:
```diff
@@ -33,7 +33,10 @@ module.exports = class Core {
       this._fetchAvailableStudies()
           .then(studies => this._updateInstalledStudies(studies));
-
+            // Register a listener to react to remote-settings updates.
+            browser.firefoxPrivilegedApi.onRemoteSettingsSync.addListener((result) => {
+              this._sendStateUpdateToUI(result[0].data);
+          });
```
nit: the indentation looks odd here :-)

rhelmer replied:
Hm yes my vscode eslint seems to disagree with ours an I haven't figured out why yet, so I have not been using auto-format :( Will figure this out.

rhelmer replied:
Oh @Dexterp37 so as I mentioned in a separate comment, I'm pretty sure I'm passing the right data here (the same as `_fetchAvailableStudies()` returns) but I don't see the UI update, I have been focusing on getting all the RS bits right but I haven't figured this out yet.

Dexterp37 commented on core-addon/FirefoxPrivilegedApi.js:
```diff
@@ -4,17 +4,24 @@
+  "resource://services-settings/remote-settings.js"
+);
+
+const { TelemetryController } = ChromeUtils.import(
+  "resource://gre/modules/TelemetryController.jsm"
+);
+
```
Why was this moved? (Was linting complaining or something?) - Nothing to do, but I'm curious :)

rhelmer replied:
Just putting them in lexical order :) I wonder if there's an eslint rule for that.

rhelmer replied:
I guess by that logic `XPCOMUtils` should be moved below but it's a common practice to have it be the first so it can be used for lazy imports (which I will likely switch these to)

Dexterp37 replied:
Makes sense!

Dexterp37 commented on core-addon/FirefoxPrivilegedApi.js:
```diff
@@ -4,17 +4,24 @@
   gUUIDGenerator: ["@mozilla.org/uuid-generator;1", "nsIUUIDGenerator"],
 });
 
+const STUDY_COLLECTION_KEY = "rally-studies-v1";
```
nit: can we add a one line comment explaining what's this for?

rhelmer replied:
I went a little overboard but I think it's probably appropriate :)

Dexterp37 commented on core-addon/FirefoxPrivilegedApi.schema.json:
```diff
@@ -2,6 +2,13 @@
   {
     "namespace": "firefoxPrivilegedApi",
     "description": "experimental API extensions for exposing privileged Firefox APIs",
+    "events": [
```
ah! TIL! I'm not an expert of this privilege API exposure. Let me bring in somebody else :)

Dexterp37 replied:
Hey @rpl, would you kindly review the changes to this "experimental API"?

rhelmer replied:
Thanks :) FTR the `EventManager` thing is great :)

Dexterp37 commented on core-addon/FirefoxPrivilegedApi.js:
```diff
@@ -32,8 +39,30 @@ this.firefoxPrivilegedApi = class extends ExtensionAPI {
-    }
+        },
+        async getRemoteSettings() {
+          return RemoteSettings(STUDY_COLLECTION_KEY).get();
```
Can `RemoteSettings(STUDY_COLLECTION_KEY)` ever fail/return `undefined` or `null`?

rhelmer replied:
It should always return `[]` I believe, if there's any problem, but I'll double-check this.

leplatrem replied:
There is not global `try/finally` on the method, but yes, the returned value should always be a list ([source](https://searchfox.org/mozilla-central/rev/69babd862de70cabfa1d0a369d38e4881bd41e4d/services/settings/RemoteSettingsClient.jsm#343))

Dexterp37 commented on core-addon/Core.js:
```diff
@@ -498,8 +501,9 @@ module.exports = class Core {
-      const request = await fetch(this._userArguments.availableStudiesURI);
-      return (await request.json()).data;
+      const result = await browser.firefoxPrivilegedApi.getRemoteSettings();
+      const studies = result[0].data;
```
Do we need some defensive programming here? Is `result` guaranteed to be a valid array that has at least an item with the `data` property?

rhelmer replied:
Yes, this is definitely a "WIP" bit :) I'd like to unpack the datastructure in a more sane way. I may move this to the privileged API implementation, so it returns something that's more useful to us.

For `about:ion` we only care about maintaining a single collection (not a set of collections in a record which is quite common), so we could do something much simpler here.

rhelmer replied:
We could consider something like having a separate record for each study, but I'm not sure that's helpful since we really want to be able to ensure that a given set of studies are interoperable, it just seems like more overhead and work to have to do.

It's essentially the difference between maintaining one single large JSON file vs. a bunch of small independent JSON files, the latter feels a little cleaner somehow but I think it'd just be overhead with no upside in our case.

Dexterp37 replied:
> It's essentially the difference between maintaining one single large JSON file vs. a bunch of small independent JSON files, the latter feels a little cleaner somehow but I think it'd just be overhead with no upside in our case.

I agree!

rhelmer replied:
I may have mentioned this elsewhere, but I figured this out and it was having an incorrect document that we've been testing against (it includes the "envelope" from RS which we probably don't want, but that's what you get when you pull directly from the server via `fetch()`)

We should have a JSON schema and check against it IMHO, both for development and also for production.

rhelmer replied:
As I was implementing this, I had to refresh my memory about how we use RS in `about:ion`:

1. we *are* using a set of records in the collection, one for each study
2. the `RemoteSettings(KEY).get()` API returns all records together as an `Array` in JS

I reviewed the production config and set up the dev server correctly, and the current changeset seems to be working, including updating when RS does.

Dexterp37 reviewed (commented):
This approach looks good to me! I'd just love other experts to validate:

- our usage of the `RemoteSettings`
- the exposed privileged API (someone that works with addons API, maybe :rpl/Luca Greco)

> 
> 
> @Dexterp37 OK so this works (well outside of the listener part but I am confident we can figure that out).

What's wrong with the listener?

> However, looking at this now it occurs to me that we could do this in unprivileged code by just fetching the appropriate JSON directly from the Remote Settings CDN and implement the listener to fire on a timer that only runs when the user is idle.
> 
> The overhead of using the built-in Firefox API doesn't seem worth it, the only advantage is really that it supports push but that doesn't really seem very compelling (some clients can't use push and fall back to an RS timer, anyway).
> 
> @Dexterp37 wdyt?

I agree on this second point. However, there might be other edge cases/trickeries that the Firefox implementation subtly takes care of (for us, thankfully!). An unprivileged API is definitely a blocker for future expansions. What about we keep the current implementation, fix the tests/listeners and file a follow-up issue to re-evaluate once we _require_ this new step?

rhelmer commented:
> This approach looks good to me! I'd just love other experts to validate:
> 
>     * our usage of the `RemoteSettings`
> 
>     * the exposed privileged API (someone that works with addons API, maybe :rpl/Luca Greco)
> 
> 
> > @Dexterp37 OK so this works (well outside of the listener part but I am confident we can figure that out).
> 
> What's wrong with the listener?

Sorry, the listener itself is fine and fires when I run `RemoteSettings.pollChanges()` as expected, but the UI does not update.

> > However, looking at this now it occurs to me that we could do this in unprivileged code by just fetching the appropriate JSON directly from the Remote Settings CDN and implement the listener to fire on a timer that only runs when the user is idle.
> > The overhead of using the built-in Firefox API doesn't seem worth it, the only advantage is really that it supports push but that doesn't really seem very compelling (some clients can't use push and fall back to an RS timer, anyway).
> > @Dexterp37 wdyt?
> 
> I agree on this second point. However, there might be other edge cases/trickeries that the Firefox implementation subtly takes care of (for us, thankfully!). An unprivileged API is definitely a blocker for future expansions. What about we keep the current implementation, fix the tests/listeners and file a follow-up issue to re-evaluate once we _require_ this new step?

WFM! The good news is that it looks pretty straightforward either way for the implementation, test coverage is what I am figuring out now.

I guess one concern I have is if we're leaving some kind of leftover trace in the Firefox `RemoteSettings` internal storage, I'll ask @leplatrem about that part specifically.

rhelmer commented:
> I guess one concern I have is if we're leaving some kind of leftover trace in the Firefox `RemoteSettings` internal storage, I'll ask @leplatrem about that part specifically.

After going for a walk and giving this more though :) so it's not usually possible for webextensions to reliably run code on shutdown, we should be able to do so for privileged extensions. So if there's any clean up to do we can do it there.

However! There is one edge case which we may or may not want to spend a lot of time on:

1. Rally core add-on is installed
2. Firefox is shut down
3. Rally add-on `.xpi` is removed from the filesystem
4. Firefox starts up and registers that the extension is not installed

Since the code is removed entirely in (3) Firefox is not able to run any sort of shutdown code in (4). If we're using pure webextension APIs and storage and such Firefox handles this for us in (4) but the privileged code is simply gone so there isn't a great way to do that.

In any case it might not be an issue here (it's possible there's no cleanup needed for the internal RS) but this is something we should verify.

rhelmer commented:
> I agree on this second point. However, there might be other edge cases/trickeries that the Firefox implementation subtly takes care of (for us, thankfully!). An unprivileged API is definitely a blocker for future expansions. What about we keep the current implementation, fix the tests/listeners and file a follow-up issue to re-evaluate once we _require_ this new step?

Oh shoot actually you know what, I missed a very important detail here which is content signing. The built-in Firefox RS API ships with a key and does verification of the content this way, which we also want (otherwise a MITM could serve a false list of studies as official).

So yes the pure-content is not as simple as it appeared :) We should consider making a library for this or reviving one of the current ones, but let's not rabbit-hole on this and move forward with the privileged approach as you've suggested.

Dexterp37 commented on core-addon/Core.js:
```diff
@@ -30,13 +30,13 @@ module.exports = class Core {
+      this._updateInstalledStudies(studies)
+    );
+    // Register a listener to react to remote-settings updates.
+    browser.firefoxPrivilegedApi.onRemoteSettingsSync.addListener((result) => {
+      this._sendStateUpdateToUI(result[0].data);
+    });
```
This is not working because `_sendStateUpdateToUI` doesn't take an argument. To make responding to changes work, you actually need to do something like this:

```suggestion
    browser.firefoxPrivilegedApi.onRemoteSettingsSync.addListener(async function (result) {
            // Important: this may be racing with the initial update or if there's two consecutive updates?
            this._availableStudies = await this._updateInstalledStudies(result[0].data);
            this._sendStateUpdateToUI();
    });
```

rhelmer replied:
aha, thank you! I figured I was just doing something wrong.

rhelmer replied:
(well I can't take this suggestion as-is since this is a constructor so we can't `await` but I understand what you mean :) )

rhelmer replied:
Ugh OK well you are using an `async` function for the callback, nevermind! `await` sounds great.

leplatrem commented on core-addon/FirefoxPrivilegedApi.js:
```diff
@@ -26,14 +39,41 @@ this.firefoxPrivilegedApi = class extends ExtensionAPI {
+            });
+            return () => {
+              // eslint-disable-next-line no-undef
+              RemoteSettings(STUDY_COLLECTION_KEY).off("sync");
```
I just realize that `.off()` expects a callback reference :( https://searchfox.org/mozilla-central/rev/69babd862de70cabfa1d0a369d38e4881bd41e4d/services/settings/RemoteSettingsClient.jsm#114

leplatrem reviewed (changes requested):
> I guess one concern I have is if we're leaving some kind of leftover trace in the Firefox RemoteSettings internal storage, I'll ask @leplatrem about that part specifically.

Leaving traces in Chrome IDB is not really an issue. Data is indexed by collection name. The only impact is that the client will continue to sync the data from the server.

If that's an issue, the collection data can be cleared with  `await RemoteSettings(STUDY_COLLECTION_KEY).db.clear()`

rhelmer commented:
Hrm OK so one thing I didn't anticipate that touches existing code is that https://github.com/mozilla-extensions/rally-core-addon/blob/master/public/locally-available-studies.json is what we've been testing against and is the direct result of `fetch()`ing from the RS server, but I think we want to make things a little nicer and have `firefoxPrivilegedApi.getRemoteSettings()` and the listener just pass an array. This is why we had to unpack the data structure more than it seemed like was necessary.

I'll clean this up now in a separate commit, I think it's only going to get more confusing if we try to work around it.

rpl reviewed (commented):
@Dexterp37 follows a couple of comments, I forwarded a more complete review pass on this to @mixedpuppy (especially because he did review experimental apis more often than me in the past and he may spot things that I may miss)

rhelmer commented on core-addon/FirefoxPrivilegedApi.js:
```diff
@@ -4,17 +4,30 @@
   "resource://gre/modules/XPCOMUtils.jsm"
 );
 
+// FIXME we shouldn't be failing no-undef eslint here, maybe we are not
+//       importing all the Firefox eslint rules properly?
```
Ah yeah looks like we're just going the `globals` approach:
https://github.com/mozilla-rally/core-addon/blob/master/.eslintrc.js#L17

I'll just do that here then thanks :)

rhelmer replied:
Hm so that's a little frustrating, something like:
```
const { XPCOMUtils } = ChromeUtils.import(
  "resource://gre/modules/XPCOMUtils.jsm"
);
```

works OK because it's defining that new `const` so eslint is cool with it (since `ChromeUtils` is allowed as a global) but of course doing it this way:

```
XPCOMUtils.defineLazyModuleGetters(this, {
  RemoteSettings: "resource://services-settings/remote-settings.js",
  TelemetryController: "resource://gre/modules/TelemetryController.jsm",
});
```

Means that `RemoteSettings` and `TelemetryController` and anything else we import would need to be added to the allowed globals, which seems... not great :) So it might make more sense to just use comments, so we don't bloat the eslint config.

rhelmer commented:
This seems to be working overall (including updating real-time as soon as RS does!) except that while the core add-on does list all available studies in the UI, it not seem to trust the list of study add-ons anymore for some reason. I'll track that down in the morning.

Once that's figured out, I think all that's left are tests. There were a few things that I will split out to separate issues:

1. joining Rally fails if there are no available studies (if RS returns `[]`, which is expected upon error)
2. we should have a shared schema with the server, and check against it at test- and run-time.

rhelmer commented:
OK finally have a green build :) I've added unit and integration tests (and refactored the integration tests some), but I am cheating and using the RS dev server ;) I'm going to figure out how to do this locally, I don't think we need to run an actual RS server though just something that mocks the API well enough for integration tests is probably fine.

Dexterp37 commented:
> 
> 
> OK finally have a green build :) I've added unit and integration tests (and refactored the integration tests some), but I am cheating and using the RS dev server ;) I'm going to figure out how to do this locally, I don't think we need to run an actual RS server though just something that mocks the API well enough for integration tests is probably fine.

Yeah, that's a good intuition. There's a couple of strategies used in m-c.

1. Directly add data to the RemoteSettings DB ([Normandy Example](https://searchfox.org/mozilla-central/rev/bd92b9b4a3c2ff022e830c1358968a84e6e69c95/toolkit/components/normandy/test/browser/browser_RecipeRunner.js#292-306), [Firefox Monitor Example](https://searchfox.org/mozilla-central/rev/bd92b9b4a3c2ff022e830c1358968a84e6e69c95/browser/components/fxmonitor/test/browser/browser_fxmonitor_doorhanger.js#57-69)).
2. Using [sinon JS to stub the return values](https://searchfox.org/mozilla-central/rev/bd92b9b4a3c2ff022e830c1358968a84e6e69c95/browser/components/extensions/test/xpcshell/test_ext_settings_overrides_defaults.js#131-142).

rhelmer commented:
> > OK finally have a green build :) I've added unit and integration tests (and refactored the integration tests some), but I am cheating and using the RS dev server ;) I'm going to figure out how to do this locally, I don't think we need to run an actual RS server though just something that mocks the API well enough for integration tests is probably fine.
> 
> Yeah, that's a good intuition. There's a couple of strategies used in m-c.
> 
>     1. Directly add data to the RemoteSettings DB ([Normandy Example](https://searchfox.org/mozilla-central/rev/bd92b9b4a3c2ff022e830c1358968a84e6e69c95/toolkit/components/normandy/test/browser/browser_RecipeRunner.js#292-306), [Firefox Monitor Example](https://searchfox.org/mozilla-central/rev/bd92b9b4a3c2ff022e830c1358968a84e6e69c95/browser/components/fxmonitor/test/browser/browser_fxmonitor_doorhanger.js#57-69)).
> 
>     2. Using [sinon JS to stub the return values](https://searchfox.org/mozilla-central/rev/bd92b9b4a3c2ff022e830c1358968a84e6e69c95/browser/components/extensions/test/xpcshell/test_ext_settings_overrides_defaults.js#131-142).
> 
> 
> I don't have any strong opinion on which technique to use. I'd say I'd slightly lean more towards 1, but we're in gut-feelings land.

Nice tip, thanks! I just used the `RemoteSettings`[...]`.db`method, it's pretty much a drop-in replacement for my "local server" method.

rhelmer commented:
@Dexterp37 I think this is pretty close, do you see any major issues? I feel like we should defer setting up a schema and simplifying the RS payload to a separate issue, in the interest of getting this landed soon. wdyt?

Dexterp37 commented on README.md:
```diff
@@ -87,6 +87,34 @@ To test the QA-signed extension in Firefox, you must be running the Nightly rele
+## Testing Remote Settings changes
+
+The Rally core-addon uses the Firefox [remote settings server](https://remote-settings.readthedocs.io/en/latest/) to publish metadata about current studies.
+Changes may be tested locally, or on the dev and staging servers, before pushing live.
```
What's needed for testing locally? Can any HTTP server work (by just serving the JSON file)?

rhelmer replied:
Hmm maybe... I was able to get it working for tests at least :) Honestly it's probably easiest to use the dev server I can link to https://remote-settings.readthedocs.io/en/latest/tutorial-dev-server.html and maybe https://remote-settings.readthedocs.io/en/latest/tutorial-local-server.html for setting up a local server.

rhelmer replied:
Although, as we discussed in other comment threads here, if we allow the `build-local-addon` feature to continue working then in general for folks building locally in dev mode it'd be easiest to use that method.

Using an actual RS server is probably more for people doing QA on builds, or testing something very specific to our RS integration.

Dexterp37 commented on core-addon/Core.js:
```diff
@@ -30,10 +30,16 @@ module.exports = class Core {
 
+    // Register a listener to react to remote-settings updates.
+    browser.firefoxPrivilegedApi.onRemoteSettingsSync.addListener((studies) => {
+      // FIXME Important: this may be racing with the initial update or if there's two consecutive updates?
```
Would you kindly file an issue about this and reference it here?

Dexterp37 commented on package.json:
```diff
@@ -15,7 +15,7 @@
     "test": "npm run --silent skip-taskcluster || (npm run test-addon && npm run test-support-library)",
     "test-addon": "npm run --silent skip-taskcluster || (mocha --require  \"./tests/hooks.js\" \"./tests/core-addon/unit/*.js\")",
-    "test-integration": "npm run --silent skip-taskcluster || (npm run build && mocha --timeout 30000 \"./tests/core-addon/integration/*.js\")",
+    "test-integration": "npm run --silent skip-taskcluster || (npm run build && mocha --unhandled-rejections=strict --timeout 30000 \"./tests/core-addon/integration/*.js\")",
```
Ah, nice. If I understand this correctly, tests should fail earlier with this flag in case of problems, without being falsely marked as positive. Is that correct? 

Would you kindly file an issue for adding this mode to the other test commands as well?

rhelmer replied:
Sure! Yes essentially it just errors out if any promises reject and it's never handled by anything, I noticed this when I forgot to do an `await` but the tests seemed to pass anyway (it was a legit error in the test),

rhelmer replied:
issue #317

Dexterp37 commented on public/locally-available-studies.json:
```diff
@@ -1,37 +1,35 @@
-    }
-  ]
-}
+[
```
Does this mean the `npm run build-local-addon` command would still work after these changes?

rhelmer replied:
Hm, the way this is currently implemented I think remote-settings is always used... per my comment above I will just go ahead and implement the old behavior in dev mode so it can still work, but we're going to need to be careful to build our production XPIs properly. Shouldn't be a big deal.

Dexterp37 commented on package.json:
```diff
@@ -15,7 +15,7 @@
     "storybook": "start-storybook -p 6006 -s ./",
```
Would you kindle file an issue about changing the `build` command to not use `--config-study-list-url` anymore? Our final releases should point to the remote settings server and the local list should exclusively be used for local testing.

We might want to keep this around for some more time to have an end-to-end QA loop.

rhelmer replied:
Oh shoot, I forgot to deal cleaning up around the local list, thanks... I've given this some thought and actually I'm not sure if it will reduce or increase confusion having multiple ways to do this. I can make the local list continue working in the meantime while we figure this out.

Dexterp37 commented on public/locally-available-studies.json:
```diff
@@ -1,37 +1,35 @@
+    },
+    "isDefault": false,
+    "sourceURI": {
+      "spec": "https://addons.mozilla.org/firefox/downloads/file/3707495/ion_basic_study-0.3.1-fx.xpi"
```
Since we're changing this file, what about changing this link to `https://addons.mozilla.org/firefox/downloads/latest/ion-basic-study` so that it always downloads the latest version? (this way we don't have to cut a new release of the core addon for template changes while we are in QA mode..)
```suggestion
      "spec": "https://addons.mozilla.org/firefox/downloads/latest/ion-basic-study"
```

rhelmer replied:
We can, the problem though is that the `version` field will then be wrong... that field isn't really used for anything right now but it's a consideration.

Dexterp37 commented on tests/core-addon/integration/core_addon.js:
```diff
@@ -28,63 +28,154 @@ async function findAndAct(driver, element, action) {
-      WAIT_FOR_PROPERTY
-    );
+// A default set of test studies.
+// FIXME - this currently loads data from remote! We need to figure out how to host this locally.
```
What data is loaded remotely? Do you mean the basic study?

rhelmer replied:
Yes, the local study and the icons. Everything with an `https://` link in here needs to be changed to either hit a local server, or we can see if `file://` works. I think it probably does.

rhelmer replied:
This is basically https://github.com/mozilla-rally/rally-core-addon/issues/253 which I think needs a followup issue?

rhelmer replied:
https://github.com/mozilla-rally/rally-core-addon/issues/319

Dexterp37 commented on tests/core-addon/integration/core_addon.js:
```diff
@@ -28,63 +28,154 @@ async function findAndAct(driver, element, action) {
+// A default set of test studies.
+// FIXME - this currently loads data from remote! We need to figure out how to host this locally.
+//         `file://` URLs might work, but we may need to resort to using a local webserver.
+const rallyTestStudy = {
```
Is there any value in re-using the json file we expose in `public/`?

rhelmer replied:
Yes and I had intended to do so thanks :)

Dexterp37 reviewed (commented):
This looks really good overall! I left a few comments below, they are all minor. I'm happy to take another look after they're addressed, I think we might merge shortly after that.

Dexterp37 reviewed (changes requested):
This looks good overall, there's a few things that need to be addressed and small nits here and there. I'll make sure to review this one last time and give the final r+ today!

rhelmer commented on core-addon/Core.js:
```diff
@@ -11,30 +11,43 @@ const OPTIONS_PAGE_PATH = "public/index.html";
+   */
   constructor(args = {}) {
-    this._userArguments = {...DEFAULT_ARGS, ...args};
+    this._userArguments = { ...DEFAULT_ARGS, ...args };
```
https://github.com/mozilla-rally/rally-core-addon/issues/327

rhelmer commented on core-addon/Core.js:
```diff
@@ -179,7 +192,7 @@ module.exports = class Core {
       case "enrollment":
         return this._enroll()
-                   .then(r => this._sendStateUpdateToUI());
+          .then(r => this._sendStateUpdateToUI());
```
https://github.com/mozilla-rally/rally-core-addon/issues/327

rhelmer commented on tests/core-addon/integration/core_addon.js:
```diff
@@ -28,63 +29,121 @@ async function findAndAct(driver, element, action) {
+ * focused on the options page.
+ */
+async function installRally(driver) {
+  this.driver = driver;
```
I did this initially so it'd be easier to refactor without a bunch of little changes, but yeah I can go ahead and do that.

rhelmer commented on tests/core-addon/integration/core_addon.js:
```diff
@@ -28,63 +29,121 @@ async function findAndAct(driver, element, action) {
-    const newTab = (await this.driver.getAllWindowHandles())[1];
-    await this.driver.switchTo().window(newTab);
+  // Switch back to web content context.
+  await this.driver.setContext(firefox.Context.CONTENT);
```
Yes, as long as we leave it set to whatever we've told the caller they can expect in the jsdoc is fine :)

Dexterp37 reviewed (approved):
r+ with the changelog file adjusted (the added line needs to be moved under "unreleased", automerge added it in 0.8)
